### PR TITLE
capture consensus discussion of 10/10/2018

### LINKED ISSUE #445: [RFC] Consensus
RFC is [here](https://github.com/solana-labs/solana/blob/master/rfcs/rfc-002-consensus.md) (afer #439 is merged)

@bhk @rob-solana 

RFC PRs:
#439

### PR DESCRIPTION
**Pull request by rob-solana**
Fixes #445

### CODE REVIEW

File: rfcs/rfc-002-consensus.md [hunk 1]
```diff
@@ -2,27 +2,31 @@
 
 VERY WIP
```
Comment (garious) [inline rfcs/rfc-002-consensus.md]: Delete this

**Comment by rob-solana** [comment]
reviewers: I have a script running that keeps the current commit up to date with what I'm typing

File: rfcs/rfc-002-consensus.md [hunk 2]
```diff
@@ -75,48 +82,101 @@ When the `Slash` vote is processed, validators should lookup `PoH hash` at `PoH
...
+The goal of this is to discourage leaders from generating multiple PoH streams.  When this occurs, and a leader is slashed, the network adopts ticks for that leader's period.
 
 ## Validation Vote Contract
 
 The goal of this contract is to simulate economic cost of mining on a shorter branch.
 
-1. With my signature I am certifying that I computed `state hash` at `PoH count` and `PoH hash`.
-2. I will not vote on a branch that doesn't contain this message for at least `N` counts, or until `PoH count` + `N` is reached by the PoH stream.
+1. With my signature I am certifying that I computed `state hash` at `PoH count tick count` and `PoH hash`.
+2. I will not vote on a branch that doesn't contain this message for at least `N` counts, or until `PoH tick count` + `N` is reached by the PoH stream.
```
Comment (aeyakovenko) [inline rfcs/rfc-002-consensus.md]: @rob-solana also double my previous lockups on this branch.

File: rfcs/rfc-002-consensus.md [hunk 3]
```diff
@@ -75,48 +82,101 @@ When the `Slash` vote is processed, validators should lookup `PoH hash` at `PoH
...
+
+Leader's transmit for a count of `T` PoH ticks.  When `T` is reached all the validators should switch to the next ranked leader.  To rank leaders, the supermajority + `M` nodes are shuffled using the above calculated random seed.
+
+All `T` ticks must be observed from the current leader for that part of PoH to be accepted by the network.  If `T` ticks (and any intervening transactions) are not observed, the network optimistically fills in the `T` ticks, and continues with PoH from the next leader.
+
+## Partitions, Forks
+
+Forks may arise at PoH tick counts that correspond to leader rotations, i.e. voting points.
+
+A fork arises when a validator observes a timeout for the designated leader during that leader's period.  The validator speculatively rolls back to the beginning of that leader's period and inserts a fork of empty ticks for the period.  These empty ticks are generated by all nodes in the network at a network-specified rate for hashes/per/tick `Z`.
```
Comment (aeyakovenko) [inline rfcs/rfc-002-consensus.md]: @rob-solana not a timeout, but any failure

File: rfcs/rfc-002-consensus.md [hunk 4]
```diff
@@ -75,48 +82,101 @@ When the `Slash` vote is processed, validators should lookup `PoH hash` at `PoH
...
+
+There are only two possible versions of the PoH during a voting period: PoH with `T` ticks and entries generated by the current leader, or PoH with just ticks.  The "just ticks" version of the PoH can be thought of as a virtual ledger, one that all nodes in the network can derive from the last tick in the previous period.
```
Comment (aeyakovenko) [inline rfcs/rfc-002-consensus.md]: there could be more forks, basically any transactional fork that wasn't finalized by supermajority could produce a different fork

**Comment by rob-solana** [comment]
I've made more updates, would appreciate feedback.

**Comment by garious** [comment]
This is very high quality. I'm going to merge it and see what I can do to bring the rest of the RFCs up to this level.

**Pull request merged**